apply_factors_wrangler.py

```diff
@@ -149,22 +151,33 @@ def lambda_handler(event, context):
         dropped_rows = non_responder_dataframe_with_prev[
             ~non_responder_dataframe_with_prev[reference].isin(
                 non_responders_with_factors[reference])].dropna()
-        if (len(dropped_rows) > 0):
-            merge_values = distinct_values.remove(region_column)
-            if merge_values is None:
-                regionless_factors = \
-                    factors_dataframe[
-                        produce_columns("imputation_factor_",
-                                        questions_list,
-                                        distinct_values)
-                    ][factors_dataframe[region_column] == regionless_code]
-
+        if len(dropped_rows) > 0:
+            merge_values = distinct_values
+            merge_values.remove(region_column)
+
+            regionless_factors = \
+                factors_dataframe[
+                    produce_columns("imputation_factor_",
+                                    questions_list,
+                                    distinct_values)
+                ][factors_dataframe[region_column] == regionless_code]
+
+            if merge_values is not None:
                 dropped_rows_with_factors = \
                     pd.merge(dropped_rows, regionless_factors,
                              on=merge_values, how="inner")
-                non_responders_with_factors = \
-                    pd.concat([non_responders_with_factors, dropped_rows_with_factors])
-                logger.info("Successfully merged missing rows with non_responders")
+            else:
+                dropped_rows["Temp_Key"] = 0
+                regionless_factors["Temp_Key"] = 0
+                dropped_rows_with_factors = \
+                    pd.merge(dropped_rows, regionless_factors,
```

Comment:
Whilst it should always be 1 row at this point, might it be worth adding something to make sure. Cause if there were more than one row it'd get large fast.

Reply:
I checked the output of calculate imputation factors. It drops duplicates so if while adding more here might be good, I'd prefer adding something to imp factors so it couldn't write out more.